webapp/channels/src/components/admin_console/search_keyword_marking.tsx

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import debounce from 'lodash/debounce';
+import Mark from 'mark.js';
+import React, {useEffect, useMemo, useRef} from 'react';
+import type {ReactNode} from 'react';
+
+type Props = {
+    keyword?: string;
+    pathname?: string;
+    children: ReactNode;
+}
+
+const DEBOUNCE_WAIT_TIME = 300;
+
+function SearchKeywordMarking(props: Props) {
+    const containerRef = useRef<HTMLDivElement>(null);
+    const markJsRef = useRef<Mark>();
+
+    function doMark(keyword: string) {
+        markJsRef.current = new Mark(containerRef.current as HTMLDivElement);
+        markJsRef.current.mark(keyword, {
+            accuracy: 'complementary',
+            exclude: ['.ignore-marking *'],
+        });
+    }
+
+    const debouncedRedrawHighlight = useMemo(() => debounce(() => {
+        if (!props.keyword || !containerRef.current) {
+            return;
+        }
+
+        if (markJsRef.current) {
+            // We need to mark again only after its 'done' callback is called
+            markJsRef.current.unmark({done: () => doMark(props.keyword as string)});
+        } else {
+            // If there's no previous instance, just create a new one
+            doMark(props.keyword);
+        }
+    }, DEBOUNCE_WAIT_TIME), [props.keyword]);
```

Comment:
I don't think `useMemo` works with the debounce since it'll create a new copy of the debounced function every time the keyword changes. I think it might work how you want it if you store the debounced function in a `useRef` and pass in the new keyword as an argument to that like

```js

const redrawHighlight = useRef(debounce((keyword) => {
    // ...
}));

useEffect(() => {
    debouncedRedrawHighlight.current(props.keyword);

    // ...
}, [props.keyword, props.pathname]);
```

That way, there'll only ever be one copy of the function and we'll just keep calling it with the latest value of `keyword`

Reply:
i redid the implementation with correct dependencies